### PR TITLE
feat: support async crop

### LINKED ISSUE #161: Add asynchronous crop support

### PR DESCRIPTION
**Pull request by guoyingtao**
fix #161

### CODE REVIEW

File: Sources/Mantis/CropView/CropView.swift [hunk 1]
```diff
@@ -700,6 +717,19 @@ extension CropView {
...
+    
+    /// completion is called in the main thread
+    func asynCrop(completion: @escaping (_ cropOutput: CropOutput) -> Void ) {
+        activityIndicator.isHidden = false
+        activityIndicator.startAnimating()
+        
+        DispatchQueue.global(qos: .userInteractive).async {
+            let cropOutput = self.crop()
+            DispatchQueue.main.async {
+                completion(cropOutput)
```
Comment (vanniktech) [inline Sources/Mantis/CropView/CropView.swift:729]: Don't you need to hide the `activityIndicator`?

**Review by vanniktech — COMMENTED** [review]
Would be nice to also have an API where `activityIndicator` isn't used at all. I'd want to have my own loading view shown instead of the one from the library

File: Sources/Mantis/CropView/CropView.swift (hunk 1 above)
Comment (guoyingtao) [reply Sources/Mantis/CropView/CropView.swift:729]: I will fix it

**Comment by guoyingtao** [comment]
> Would be nice to also have an API where `activityIndicator` isn't used at all. I'd want to have my own loading view shown instead of the one from the library

Thanks for the suggestion! I will think about it.

File: Sources/Mantis/CropView/CropView.swift [hunk 2]
```diff
@@ -700,6 +717,20 @@ extension CropView {
     func crop() -> CropOutput {
         return crop(image)
     }
+    
+    /// completion is called in the main thread
+    func asyncCrop(completion: @escaping (_ cropOutput: CropOutput) -> Void ) {
+        activityIndicator.isHidden = false
+        activityIndicator.startAnimating()
+        
+        DispatchQueue.global(qos: .userInteractive).async {
```
Comment (vanniktech) [inline Sources/Mantis/CropView/CropView.swift:726]: Do we also need?

```suggestion
        DispatchQueue.global(qos: .userInteractive).async { [weak self] in
```

Comment (guoyingtao) [reply Sources/Mantis/CropView/CropView.swift:726]: No it is not needed. DispatchQueue does not hold strong self to it won't cause retain cycle.
You can test it by putting a breakpoint in `deinit`

**Comment by guoyingtao** [comment]
I found some code in crop() which should only be called in the main thread. I need to fix them.

**Comment by guoyingtao** [comment]
> I found some code in crop() which should only be called in the main thread. I need to fix them.

Fixed

**Comment by guoyingtao** [comment]
> Would be nice to also have an API where `activityIndicator` isn't used at all. I'd want to have my own loading view shown instead of the one from the library

I created a new issue #233 and plan to do it in a separate PR.

**Review by vanniktech — APPROVED** [review]

**Pull request merged**